episodes/4-advanced-layer-types.Rmd

```diff
@@ -785,6 +787,220 @@ This is called hyperparameter tuning.
+- C: Incorrect, it is important to note that you actually train and test different models for each run of the grid search
+- D: Incorrect, this is the worst model since the validation loss is highest
+- E: Correct, this is the best model with the lowest loss
+- F: Incorrect, it could be that a different number of layers in combination with a dropout rate that we did not test (for example 3 layers and a dropout rate of 0.6) could be the best model.
+
+2 (Optional): 
+
+You need to adapt the code as follows:
+```python
+def create_nn_with_hp(activation_function, kernel_size):
+    inputs = keras.Input(shape=train_images.shape[1:])
+    x = inputs
+    for layer in range(3):
+        x = keras.layers.Conv2D(50, (kernel_size, kernel_size), activation=activation_function)(x)
+        x = keras.layers.MaxPooling2D((2, 2))(x)
+    x = keras.layers.Dropout(0.2)(x)
+    x = keras.layers.Flatten()(x)
+    x = keras.layers.Dense(50, activation=activation_function)(x)
+    outputs = keras.layers.Dense(10)(x)
+    model = keras.Model(inputs=inputs, outputs=outputs, name="cifar_model")
+    return model
+    
+hp = keras_tuner.HyperParameters()
+
+def build_model(hp):
+    kernel_size = hp.Int("kernel_size", min_value=3, max_value=4, step=1)
+    activation = hp.Choice("activation", ["relu", "tanh"])
+    model = create_nn_with_hp(activation, kernel_size)
+    compile_model(model)
+    return model
+    
+tuner = keras_tuner.GridSearch(build_model, objective='val_loss')
+tuner.search(train_images, train_labels, epochs=20,
+             validation_data=(val_images, val_labels))
+```
+```output
+Trial 4 Complete [00h 00m 56s]
+val_loss: 1.44037926197052
+
+Best val_loss So Far: 1.2700632810592651
+Total elapsed time: 00h 04m 01s
+```
+Let's look at the results:
+```python
+tuner.results_summary()
+```
+```output
+Results summary
+Results in ./untitled_project
+Showing 10 best trials
+Objective(name="val_loss", direction="min")
+
+Trial 0000 summary
+Hyperparameters:
+kernel_size: 3
+activation: relu
+Score: 1.2700632810592651
+
+Trial 0001 summary
+Hyperparameters:
+kernel_size: 3
+activation: tanh
+Score: 1.2945374250411987
+
+Trial 0002 summary
+Hyperparameters:
+kernel_size: 4
+activation: relu
+Score: 1.431167483329773
+
+Trial 0003 summary
+Hyperparameters:
+kernel_size: 4
+activation: tanh
+Score: 1.44037926197052
+```
+A kernel size of 3 and `relu` as activation function is the best tested combination.
+
+::::
+::: 
+
+Grid search can quickly result in a combinatorial explosion because all combinations of hyperparameters are trained and tested.
+Instead, `random search` randomly samples combinations of hyperparemeters, allowing for a much larger look through a large number of possible hyperparameter combinations.
+
+Next to grid search and random search there are many different hyperparameter tuning strategies, including `neural architecture search` where a separate neural network is trained to find the best architecture for a model!
```

Comment:
Maybe add a link (to Wikipedia?) for more information about neural architecture search?

```suggestion
Next to grid search and random search there are many different hyperparameter tuning strategies, including [neural architecture search](https://en.wikipedia.org/wiki/Neural_architecture_search) where a separate neural network is trained to find the best architecture for a model!
```